gcsio/src/test/java/com/google/cloud/hadoop/gcsio/WatchdogTest.java

```diff
@@ -0,0 +1,247 @@
+package com.google.cloud.hadoop.gcsio;
+
+import static com.google.common.truth.Truth.assertThat;
+import static org.mockito.ArgumentMatchers.anyString;
+import static org.mockito.ArgumentMatchers.isA;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.never;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+import com.google.api.core.ApiClock;
+import com.google.api.core.CurrentMillisClock;
+import com.google.common.flogger.GoogleLogger;
+import com.google.storage.v2.ReadObjectResponse;
+import com.google.storage.v2.WriteObjectRequest;
+import com.google.storage.v2.WriteObjectResponse;
+import io.grpc.ClientCall;
+import io.grpc.Context;
+import io.grpc.Context.CancellableContext;
+import io.grpc.stub.StreamObserver;
+import java.time.Duration;
+import java.util.Iterator;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeoutException;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+@RunWith(JUnit4.class)
+public class WatchdogTest {
+  private static final GoogleLogger logger = GoogleLogger.forEnclosingClass();
+
+  private Watchdog watchdog;
+
+  private final Duration waitTime = Duration.ofSeconds(5);
+  private final Duration zeroWaitTime = Duration.ofSeconds(0);
+
+  @Before
+  public void setUp() throws Exception {
+    Duration checkInterval = Duration.ofSeconds(2);
+    ApiClock clock = CurrentMillisClock.getDefaultClock();
+    ScheduledExecutorService executor = Executors.newSingleThreadScheduledExecutor();
+    watchdog = Watchdog.create(clock, checkInterval, executor);
+  }
+
+  @Test
+  public void testPassThroughClientStreamingRPC() {
+    ClientCall<WriteObjectRequest, WriteObjectResponse> clientCall = mock(ClientCall.class);
+    StreamObserver<WriteObjectRequest> streamObserver = mock(StreamObserver.class);
+    StreamObserver<WriteObjectRequest> watch = watchdog.watch(clientCall, streamObserver, waitTime);
+    WriteObjectRequest value = WriteObjectRequest.newBuilder().build();
+    watch.onNext(value);
+    verify(streamObserver).onNext(value);
+    Throwable t = mock(Throwable.class);
+    watch.onError(t);
+    verify(streamObserver).onError(t);
+    watch.onCompleted();
+    verify(streamObserver).onCompleted();
```

Comment:
Mock-based tests are in general hard to maintain and reason about, and specifically discouraged for testing gRPC: http://g3doc/net/grpc/g3doc/java/unit_test

We get rid of them in favor of fakes and stubs for virtually all the test already, may we not re-introduce them for testing gRPC?

Reply:
Done